blocksync/reactor.go

```diff
@@ -68,17 +68,35 @@ type Reactor struct {
 func NewReactor(state sm.State, blockExec *sm.BlockExecutor, store *store.BlockStore,
 	blockSync bool, metrics *Metrics,
 ) *Reactor {
-	if state.LastBlockHeight != store.Height() {
-		panic(fmt.Sprintf("state (%v) and store (%v) height mismatch", state.LastBlockHeight,
-			store.Height()))
-	}
 
+	storeHeight := store.Height()
+	var err error
+	if storeHeight == 0 {
+		// If state sync was performed offline and the stores were bootstrapped to height H
+		// the state stores lastHeight will be H while blockstores Height and Base are still 0
+		// 1. This scenario should not lead to a panic in this case, which is indicated by
+		// having a OfflineStateSyncHeight > 0
+		// 2. We need to instruct the blocksync reactor to start fetching blocks from H+1
+		// instead of 0.
+		storeHeight, err = blockExec.Store().GetOfflineStateSyncHeight()
+		if err != nil && err.Error() != "value empty" {
+			panic(fmt.Sprintf("failed to retrieve statesynced height from store %s", err))
+		}
+		err = blockExec.Store().SetOfflineStateSyncHeight(0)
+		if err != nil {
+			panic("failed to reset the offline state sync height ")
+		}
```

Comment:
For context, @jmalicevic and I discussed the case when the process crashes precisely here. In that case, when restarting the process, it will fail at line 83 above. However, we decided not to try to solved this here because:

1. This problem also exists (it always has, since introduction of state-sync) with the regular (online) bootstrapping
2. This would occur when an operator tries to start a new node from scratch (bootstrapping either online or via a snapshot from the SDK). So it would happen immediately after the operator types the command. Therefore, a good error message should suffice